src/main/java/org/spongepowered/mod/plugin/SpongeModPluginContainer.java

```diff
@@ -273,15 +280,44 @@ public void constructMod(FMLConstructionEvent event) {
             }
 
             // Add source file to classloader so we can load it.
-            ModClassLoader modClassLoader = event.getModClassLoader();
+            final ModClassLoader modClassLoader = event.getModClassLoader();
             modClassLoader.addFile(getSource());
             modClassLoader.clearNegativeCacheFor(this.candidate.getClassList());
 
-            Class<?> pluginClazz = Class.forName(this.className, true, modClassLoader);
+            final Class<?> pluginClass = Class.forName(this.className, true, modClassLoader);
+
+            final Type adapterType = (org.objectweb.asm.Type) this.descriptor.get("adapter");
+            final Class<?> adapterClass;
+            if (adapterType != null) {
+                adapterClass = Class.forName(adapterType.getClassName(), true, modClassLoader);
+            } else {
+                adapterClass = PluginAdapter.Default.class;
+            }
+
+            Injector injector = spongeInjector.getParent().createChildInjector(new PluginModule((PluginContainer) this));
+
+            final ImmutableList.Builder<Module> modules = ImmutableList.builder();
+            // Check for plugin specified modules
+            final List<Type> injectionModuleTypes = (List<Type>) this.descriptor.get("injectionModules");
+            if (injectionModuleTypes != null && injectionModuleTypes.size() > 0) {
+                for (Type injectionModuleType : injectionModuleTypes) {
+                    final Class<?> moduleClass = Class.forName(injectionModuleType.getClassName(), true, modClassLoader);
+                    modules.add((Module) moduleClass.newInstance());
+                }
+            }
+
+            final PluginAdapter adapter = adapterClass == PluginAdapter.Default.class ? DefaultPluginAdapter.INSTANCE :
+                    (PluginAdapter) adapterClass.newInstance();
```

Comment:
Injector? But i'm not sure about here

Reply:
I will create the instance with the parent injector, adapters should be reusable for multiple plugin. Although it's currently not implemented this way.